### PR TITLE
detect: add vlan.id keyword - v4

### PR DESCRIPTION
**Pull request by AkakiAlice**
Ticket: OISF#1065

Make sure these boxes are checked accordingly before submitting your Pull Request -- thank you.

## Contribution style:
- [x] I have read the contributing guide lines at
   https://docs.suricata.io/en/latest/devguide/contributing/contribution-process.html

## Our Contribution agreements:
- [x] I have signed the Open Information Security Foundation contribution agreement at
   https://suricata.io/about/contribution-agreement/ (note: this is only required once)

## Changes (if applicable):
- [ ] I have updated the User Guide (in [doc/userguide/](https://github.com/OISF/suricata/tree/304271e63a9e388412f25f0f94a1a0da4bf619d9/doc/userguide)) to reflect the changes made
- [ ] I have updated the JSON schema (in [etc/schema.json](https://github.com/OISF/suricata/blob/304271e63a9e388412f25f0f94a1a0da4bf619d9/etc/schema.json)) to reflect all logging changes
      (including schema descriptions)
- [ ] I have created a ticket at
      https://redmine.openinfosecfoundation.org/projects/suricata/issues

Link to ticket: https://redmine.openinfosecfoundation.org/issues/1065
Describe changes:
- Introduce vlan.id keyword

### Provide values to any of the below to override the defaults.

- To use an LibHTP, Suricata-Verify or Suricata-Update pull request,
  link to the pull request in the respective `_BRANCH` variable.
- Leave unused overrides blank or remove.

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2185
Previous PR= https://github.com/OISF/suricata/pull/12273

### CODE REVIEW

**Comment by catenacyber** [comment]
Thanks, what are the differences with the previous version ?

I see CI is still red, cf the check rust build that you can reproduce locally by running `cargo clippy --fix` in rust subdirectory

File: doc/userguide/rules/vlan-id.rst [hunk 1]
```diff
@@ -0,0 +1,33 @@
+Vlan.id Keyword
+===============
+
+Suricata has a ``vlan.id`` keyword that can be used in signatures to identify
+and filter network packets based on Virtual Local Area Network IDs. By default,
+it matches all layers if a packet contains multiple VLAN layers. However, if a
+specific layer is defined, it will only match that layer.
+
+vlan.id uses :ref:`unsigned 16-bit integer <rules-integer-keywords>`.
```
Comment (catenacyber) [inline doc/userguide/rules/vlan-id.rst:9]: https://github.com/OISF/suricata/pull/12273#discussion_r1884368088

> Should we also mention here the max value, and the specification that defines that?

I think it is a good idea

File: src/detect-vlan-id.c [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        PrefilterAddSids(&det_ctx->pmq, ctx->sigs_array, ctx->sigs_cnt);
+    }
+}
+
+static void PrefilterPacketVlanIdSet(PrefilterPacketHeaderValue *v, void *smctx)
+{
+    const DetectVlanIdData *a = smctx;
+    v->u8[0] = a->du16.mode;
+    v->u8[1] = a->layer;
+    v->u16[2] = a->du16.arg1;
```
Comment (catenacyber) [inline src/detect-vlan-id.c:96]: There should also be `v->u16[3] = a->du16.arg2;` (used for ranges like vlan.id:100-200)

File: src/detect-vlan-id.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+#ifndef SURICATA_DETECT_VLAN_ID_H
+#define SURICATA_DETECT_VLAN_ID_H
+
+#define ANY_VLAN_LAYER INT8_MIN
```
Comment (catenacyber) [inline src/detect-vlan-id.h:21]: This definition should be in rust

File: doc/userguide/rules/vlan-id.rst [hunk 4]
```diff
@@ -0,0 +1,33 @@
+Vlan.id Keyword
```
Comment (catenacyber) [inline doc/userguide/rules/vlan-id.rst:1]: CI is red because you have to add this file in doc/userguide/rules/index.rst

**Comment by catenacyber** [comment]
And last : Suggestion for even more expressivity : add support vlan.id: >100,all which matches only if all vlan layers match

File: rust/src/detect/vlan_id.rs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    fn test_detect_parse_vlan_id() {
+        assert_eq!(
+            detect_parse_vlan_id("300").unwrap(),
+            DetectVlanIdData {
+                du16: DetectUintData {
+                    arg1: 300,
+                    arg2: 0,
+                    mode: DetectUintMode::DetectUintModeEqual,
+                },
+                layer: i8::MIN
```
Comment (inashivb) [inline rust/src/detect/vlan_id.rs:93]: Q: Why is the default layer `i8::MIN` when a `layer < -3` is unacceptable to the parser? ref: https://github.com/OISF/suricata/pull/12285/files#diff-19017a83531856b891bc67909647f1bdbf986c4fd068b0e107c149a0158cdd11R52

**Comment by inashivb** [comment]
Has a newer revision linked above

**Pull request closed without merging**